9.5/root/usr/share/container-scripts/postgresql/common.sh

```diff
@@ -218,3 +223,149 @@ function wait_for_postgresql_master() {
     sleep 1
   done
 }
+
+
+run_pgupgrade ()
+(
+  optimized=false
+  old_raw_version=${POSTGRESQL_PREV_VERSION//\./}
+  new_raw_version=${POSTGRESQL_VERSION//\./}
+  old_pgengine=/opt/rh/rh-postgresql${old_raw_version}/root/usr/bin
+  new_pgengine=/opt/rh/rh-postgresql${new_raw_version}/root/usr/bin
+  PGDATA_new="${PGDATA}-new"
+
+  printf >&2 "\n==========  \$PGDATA upgrade: %s -> %s  ==========\n\n" \
+             "$POSTGRESQL_PREV_VERSION" \
+             "$POSTGRESQL_VERSION"
+
+  info_msg () { printf >&2 "\n===>  $*\n\n" ;}
+
+  # pg_upgrade writes logs to cwd, so go to the persistent storage first
+  cd "$HOME"/data
+
+  # disable this because of scl_source, 'set +u' just makes the code ugly
+  # anyways
+  set +u
+
+  # we need to have the old SCL enabled, otherwise the $old_pgengine is not
+  # working.  The scl_source script doesn't pay attention to non-zero exit
+  # statuses, so use 'set +e'.
+  set +e
+  source scl_source enable rh-postgresql$old_raw_version
+  set -e
+
+  case $POSTGRESQL_UPGRADE in
+    copy) # we accept this
+      ;;
+    hardlink)
+      optimized=:
+      ;;
+    *)
+      echo >&2 "Unsupported value: \$POSTGRESQL_UPGRADE=$POSTGRESQL_UPGRADE"
+      false
+      ;;
+  esac
+
+  # Ensure $PGDATA_new doesn't exist yet, so we can immediately remove it if
+  # there's some problem.
+  test ! -e "$PGDATA_new"
+
+  # initialize the database
+  info_msg "Initialize new data directory; we will migrate to that."
+  initdb_cmd=( initdb_wrapper "$new_pgengine"/initdb "$PGDATA_new" )
+  eval "\${initdb_cmd[@]} ${POSTGRESQL_UPGRADE_INITDB_OPTIONS-}" || \
+    { rm -rf "$PGDATA_new" ; false ; }
+
+  upgrade_cmd=(
+      "$new_pgengine"/pg_upgrade
+      "--old-bindir=$old_pgengine"
+      "--new-bindir=$new_pgengine"
+      "--old-datadir=$PGDATA"
+      "--new-datadir=$PGDATA_new"
+  )
+
+  # Dangerous --link option, we loose $DATADIR if something goes wrong.
+  ! $optimized || upgrade_cmd+=(--link)
+
+  # User-specififed options for pg_upgrade.
+  eval "upgrade_cmd+=(${POSTGRESQL_UPGRADE_PGUPGRADE_OPTIONS-})"
+
+  # the upgrade
+  info_msg "Starting the pg_upgrade process."
+  "${upgrade_cmd[@]}" || { rm -rf "$PGDATA_new" && false ; }
+
+  # Move the important configuration and remove old data.  This is highly
+  # careless, but we can't do more for this over-automatized process.
+  info_msg "Swap the old and new PGDATA and cleanup."
+  mv "$PGDATA"/*.conf "$PGDATA_new"
+  rm -rf "$PGDATA"
```

Comment:
Well, I simply disagree. We start the DB right after pg_upgrade run, and there's a high risk of `ENOSPC` just because we would keep about 100% of datadir size unnecessarily.  I agree that having an additional option explicitly **asking** for not deleting the datadir might be useful ...

My point about all of this binary datadir upgrade is:  Don't do upgrades unless you really know what you are doing ... It is not place to play games;  also, I'm sure wise admins won't do upgrades unless that's really necessary. Which in turn means that they would downvote this `pg_upgrade` N->(N+1) wofkflow.